knowledge_runtime/knowledge_runtime/services/duckdb_generator.py

```
@@ -0,0 +1,709 @@
+                    "FROM information_schema.columns "
+                    "WHERE table_name = ? AND table_schema = 'main' "
+                    "ORDER BY ordinal_position",
+                    [table_name],
+                ).fetchall()
+
+                for col_name, col_type, _is_nullable in col_result:
+                    # Get null count for this column
+                    try:
+                        null_count_result = conn.execute(
+                            f'SELECT COUNT(*) FROM "{table_name}" '
+                            f'WHERE "{col_name}" IS NULL'
```

Ruff (0.15.12)

[error] Possible SQL injection vector through string-based query construction

(S608)